coMET
=====

##Overview
The coMET package is a web-based plotting tool and R-based package to visualize different genome-wide association scans such as EWAS (epigenome-wide association scan) results in a genomic region of interest. coMET provides a plot of the EWAS association signal and visualisation of the methylation correlation between CpG sites (co-methylation). The coMET package also provides the option to annotate the region using functional genomic information, including both user-defined features and pre-selected features based on the Encode project. The plot can be customized with different parameters, such as plot labels, colours, symbols, heatmap colour scheme, significance thresholds, and including reference CpG sites. Finally, the tool can also be applied to display the correlation patterns of other genomic data and any species, e.g. gene expression array data. 


##coMET webservice

The webservice is the pre-formated web service of coMET with a reduction of parameters availlable.

Go to : http://www.epigen.kcl.ac.uk/comet
or directly one of instance: http://comet.epigen.kcl.ac.uk:3838/coMET/

##Developpers

coMET is developed by Tiphaine C. Martin in collaboration with Idil Erte, Pei-Chien Tsai, Jordana T.Bell, Department of Twin Research, Kings College London.


##Cite

Martin, T.C, Erte, I, Tsai, P-C, Bell, J.T.,coMET: an R plotting package to visualize regional plots of epigenome-wide association scan results, QC14, 2014.


##Contacts

For any question, you can send an email to Tiphaine Martin (tiphaine.martin at kcl.ac.uk) and Jordana Bell (jordana.bell at kcl.ac.uk)


##More information

Go to the website Department of Twin Research :http://epigen.kcl.ac.uk/comet

Go to the help of coMET webservice : http://comet.epigen.kcl.ac.uk:3838/coMET/
