%\VignetteIndexEntry{coMET users guide}
%\VignetteDepends{coMET}
%\VignetteKeywords{Software, Visualization, GenomeAnnotation, GeneRegulation, GenomicVariation, DNAMethylation, SNP, GeneExpression }
%\VignettePackage{coMET}
%\VignetteEngine{knitr::knitr}

\documentclass[11pt]{article}

% A bunch of styles and package requirements for the Bioconductor vignette branding
<<style, eval=TRUE, echo=FALSE,results="asis">>=
#library("BiocStyle")
BiocStyle::latex()
@

<<setup, include=FALSE, cache=FALSE, echo=F>>=
library(knitr)
opts_chunk$set(fig.path='figure/minimal-', fig.align='center', fig.show='hold')
options(replace.assign=TRUE,width=90)
@

\RequirePackage[utf8]{inputenc} 
% \RequirePackage{hyperref}
\RequirePackage{url}
\RequirePackage[authoryear,round]{natbib}
\bibliographystyle{plainnat}
% \RequirePackage[text={7.2in,9in},centering]{geometry}
%\setkeys{Gin}{width=0.95\textwidth}
\RequirePackage{longtable}
\RequirePackage{graphicx}
\newcommand{\code}[1]{{\texttt{#1}}}
\newcommand{\term}[1]{{\emph{#1}}}
\newcommand{\Rmethod}[1]{{\textit{#1}}}
\newcommand{\Rfunarg}[1]{{\textit{#1}}}
\newcommand{\scscst}{\scriptscriptstyle}
\newcommand{\scst}{\scriptstyle}
\newcommand{\mgg}[0]{\Rpackage{coMET} }
\newcommand{\Reference}[1]{{\texttt{#1}}}
\newcommand{\link}[1]{{#1}}
\newcommand{\RR}[0]{{\texttt{R}}}

\title{The coMET User Guide}
\author{Tiphaine Martin \footnote{tiphaine.martin@kcl.ac.uk}, Idil Yet \footnote{idil.yet@kcl.ac.uk}, Pei-Chien Tsai \footnote{peichien.tsai@kcl.ac.uk}, Jordana T. Bell \footnote{jordana.bell@kcl.ac.uk}}
\date{Edited: September 2014; Compiled: \today}

\begin{document}

\maketitle

\section{Citation}
<<citatation>>=
citation(package='coMET')
@

\clearpage

\tableofcontents
\clearpage

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Introduction}
The CoMET package is a web-based plotting tool and R-based package to visualize EWAS (epigenome-wide association scan) results in a genomic region of interest. CoMET provides a plot of the EWAS association signal and visualisation of the methylation correlation between CpG sites (co-methylation). The CoMET package also provides the option to annotate the region using functional genomic information, including both user-defined features and pre-selected features based on the Encode project. The plot can be customized with different parameters, such as plot labels, colours, symbols, heatmap colour scheme, significance thresholds, and including reference CpG sites. Finally, the tool can also be applied to display the correlation patterns of other genomic data, e.g. gene expression array data.

coMET generates a multi-panel plot to visualize EWAS results, co-methylation patterns, and annotation tracks in a genomic region of interest. A coMET figure (cf. Fig. 1) includes three components: 
\begin{enumerate}
  \item the upper plot shows the strength and extent of EWAS association signal;
  \item the middle panel provides customized annotation tracks;
  \item the lower panel shows the correlation between selected CpG sites in the genomic region.
\end{enumerate}
The structure of the plots builds on snp.plotter (Luna et al., 2007), with extensions to incorporate genomic annotation tracks and customized functions. coMET produces plots in PDF and Encapsulated Postscript (EPS) format.


\section{Usage}
CoMET requires the installation of R, the statistical computing software, freely available for Linux, Windows, or MacOS.CoMET can be downloaded from bioconductor. Packages can be installed using the install.packages command in R.  The coMET R package includes two major functions \emph{comet.web} and \emph{comet}. the function \emph{comet.web} generates output plot with the same settings of genomic annotation tracks as that of the webservice ( http://comet.epigen.kcl.ac.uk:3838/coMET/ or http://www.epigen.kcl.ac.uk/comet). the function \emph{comet} generates output plots with the customized annotation tracks defined by user. 

<<installPackages,eval=FALSE>>=
source("http://bioconductor.org/biocLite.R")
biocLite("coMET")
@

CoMET can be loaded into R using this command:
<<dd,echo=FALSE,eval=FALSE>>=
require("hash")
require("grid")
require("grDevices")
require("biomaRt")
require("Gviz")
require("ggbio")
require("rtracklayer")
require("GenomicRanges")
require("colortools")

rdir <- system.file("R", package="coMET",mustWork=TRUE)
source(file.path(rdir, "AnalyseFile.R"))
source(file.path(rdir, "BiofeatureGraphics.R"))
source(file.path(rdir, "comet.R"))
source(file.path(rdir, "cometWeb.R"))
source(file.path(rdir, "DrawPlot.R"))
source(file.path(rdir, "GeneralMethodComet.R"))
@

<<loadPackages, eval=TRUE>>=
library(coMET)
@

The configuration file ammends the options for the plot to be produced from comet. Example configuration and input files are also provided in the webervice (http://comet.epigen.kcl.ac.uk:3838/coMET/ and http://www.epigen.kcl.ac.uk/comet). Information about the package can viewed from within R using this command.

<<findHel, eval=FALSE>>=
?comet
?comet.web
@


\section{File Formats}
4 formats (option CORMATRIX.FORMAT) describes tab-delimited correlation file 
(option CORMATRIX.FILE): 
\begin{enumerate}
\item CORMATRIX: the data has already pre-computed by users; 
\item RAW: The data is at raw format and can be computed by one of 3 methods 
Spearman, Pearson, Kendall (option CORMATRIX.METHOD). ; 
\item 2 other formats are when the association data and correlation matrix are in the 
same file defined in the option DATA.FILE (region to visualize (option DATA.FILE) 
Others describing the P-value and optionally the direction of association (for example gene expression, validation studies) (option DATA.FILE.LARGE)) 
\end{enumerate}


\subsection{Format of info file(mandatory):}
Info file can be a list of CpG sites with/without Beta value (or direction sign). If it is a site file then it is mandatory to have the 4 columns as shown below with headers in the same order. Beta can be the 5th column(optional) and it can be either a numeric value (positive or negative values) or only direction sign ("+", "-")

<<Filedata>>=
extdata <- system.file("extdata", package="coMET",mustWork=TRUE)
infofile <- file.path(extdata, "cyp1b1_infofile.txt")
#infofile <- "../inst/extdata/cyp1b1_infofile.txt" 

data_info <-read.csv(infofile, header = TRUE,
                     sep = "\t", quote = "")

head(data_info)
@

Also it can be region with/without an end start of the base pair. If it is a region file then it is mandatory to have the 5 columns below with headers in this order. Beta can be the 6th column(optional).

<<expMatrix>>=
extdata <- system.file("extdata", package="coMET",mustWork=TRUE)
infoexp <- file.path(extdata, "cyp1b1_infofile_exprGene_region.txt")
#infoexp <- "../inst/extdata/cyp1b1_infofile_exprGene_region.txt" 

data_infoexp <-read.csv(infoexp, header = TRUE,
                        sep = "\t", quote = "")

head(data_infoexp)
@

\subsection{Format of correlation matrix(mandatory):}
The data can be either pre-calculated correlation matrix or raw data. If it is a raw data then you can select the type of correlation method (spearman ,kendall or pearson).

<<corrMatrix>>=
extdata <- system.file("extdata", package="coMET",mustWork=TRUE)
corfile <- file.path(extdata, "cyp1b1_res37_rawMatrix.txt")
#corfile <- "../inst/extdata/cyp1b1_res37_rawMatrix.txt" 

data_cor <-read.csv(corfile, header = TRUE,
                    sep = "\t", quote = "")
data_cor[1:6,1:6]
@


\subsection{Format of extra info file:}

This can be another type of info file (e.g Expression data or replication data) and it follows the same rules than the info file.

\subsection{Format of annotation file}

format accepted by Gviz such as BED, GTF, and GFF3 format

\subsection{Option of config.file}

If you would like to make your own changes to the plot you can download the configuration file from this site. After you make the relative changes you can upload it to the server again and plot.

 The important options of a coMET figure includes three components: 
\begin{enumerate}
  \item the upper plot shows the strength and extent of EWAS association signal;
   \begin{itemize}
      \item PVAL.THRESHOLD : Threshold of the significance that is displayed as a red dash line
      \item DISP.ASSOCIATION : Optional works only if MYDATA.FILE= SITE\_ASSOC or REGION\_ASSOC and so the file contains the effect directions. If it is FALSE (the default), the co-methylation pattern related to reference genomic reference (defined by the option MYDATA.REF) is shown in p-value plot; if it is True, the effect directions is shown on the plot
      \item DISP.REGION : Optional works only if MYDATA.FILE= REGION or REGION\_ASSOC. If it is TRUE, the region of genomic element will be shown by a continuous line with the color of  genomic element, in addition to a symbole in center of region.
  \end{itemize}
  
  \item the middle panel provides customized annotation tracks;
     \begin{itemize}
     \item LIST.TRACKS (for \emph{comet.web} function): List of annotation tracks that is able to visualise  visualise :geneENSEMBL, CGI, ChromHMM, DNAse, RegENSEMBL, SNP, transcriptENSEMBL, SNPstoma, SNPstru, SNPstrustoma, ISCA, COSMIC, GAD, ClinVar, GeneReviews, GWAS, ClinVarCNV, GCcontent, genesUCSC, xenogenesUCSC. 
     \item TRACKS.GVIZ, TRACKS.GGBIO, TRACKS.TRACKVIEWER (for \emph{comet} function): For each option, it is possible to give a list of annotation tracks that is created by Gviz, GGBio, TrackViewer bioconductor package.
     \end{itemize}

  \item the lower panel shows the correlation between selected CpG sites in the genomic region.
    \begin{itemize}
      \item CORMATRIX.FORMAT : Format of the input CORMATRIX.FILE: either raw data (option RAW) or correlation matrix (option CORMATRIX) 
      \item CORMATRIX.METHOD : If data are raw, it will be necessary to produce the correlation matrix being able to use 3 methods (spearman, pearson and kendall).
      \item CORMATRIX.COLOR.SCHEME : There are 5 colors (heat, bluewhitered, cm, topo, gray, bluetored)
   \end{itemize}

\end{enumerate}

<<configMatrix>>=
extdata <- system.file("extdata", package="coMET",mustWork=TRUE)
configfile <- file.path(extdata, "config_cyp1b1_zoom_4webserver.txt")
#configfile <- "../inst/extdata/config_cyp1b1_zoom_4webserver.txt" 

data_config <-read.csv(configfile, quote = "")
data_config
@

\section{Creation plot like webservice:}
It is possible to use simply the creation of plot with the function comet.web, for example the figure \ref{fig:cometweb_simple}.

<<cometwebPlotText, eval=FALSE, fig.keep='last' >>=
extdata <- system.file("extdata", package="coMET",mustWork=TRUE)
configfile <- file.path(extdata, "config_cyp1b1_zoom.txt")
#configfile <- "../inst/extdata/config_cyp1b1_zoom.txt" 
comet.web(config.file=configfile,PRINT.IMAGE=FALSE,VERBOSE=FALSE)
@

\begin{figure}
<<cometwebPlot, echo=FALSE, fig.keep='last' >>=
extdata <- system.file("extdata", package="coMET",mustWork=TRUE)
configfile <- file.path(extdata, "config_cyp1b1_zoom.txt")
#configfile <- "../inst/extdata/config_cyp1b1_zoom.txt" 
comet.web(config.file=configfile,PRINT.IMAGE=FALSE,VERBOSE=FALSE)
@
\caption{Plot with comet.web function.\label{fig:cometweb_simple}}
\end{figure}

\section{Creation plot with the generic function: comet}
It is possible to create the annotation tracks by Gviz, trackviewer or ggbio.
Currently, the combinaison of annotation tracks from Gviz with the heatmap of correlation between genetics elements is better plot in coMET, for example the figure \ref{fig:cometPlot}.

\subsection{coMET plot: pvalue plot, annotation tracks, and correlation matrice}
<<cometPlotText, eval=FALSE, fig.keep='last' >>=
library(Gviz)
extdata <- system.file("extdata", package="coMET",mustWork=TRUE)
configfile <- file.path(extdata, "config_cyp1b1_zoom_4comet.txt")
#configfile <- "../inst/extdata/config_cyp1b1_zoom_4comet.txt" 
chrom <- "chr2"
start <- 38290160
end <- 38303219
gen <- "hg19"
strant <- "*"

genetrack <-genesENSEMBL(gen,chrom,start,end,showId=FALSE)
snptrack <- snpBiomart(chrom, start, end, dataset="hsapiens_snp_som",showId=FALSE)
strutrack <- structureBiomart(chrom, start, end, strand,
                              dataset="hsapiens_structvar_som",showId=FALSE)
iscatrack <-ISCATrack(gen,chrom,start,end,table="iscaPathogenic")

listgviz <- list(genetrack,snptrack,iscatrack)
comet(config.file=configfile,TRACKS.GVIZ=listgviz, 
      VERBOSE=FALSE, PRINT.IMAGE=FALSE)
@

\begin{figure}
<<cometPlot, echo=FALSE, fig.keep='last' >>=
library(Gviz)
extdata <- system.file("extdata", package="coMET",mustWork=TRUE)
configfile <- file.path(extdata, "config_cyp1b1_zoom_4comet.txt")
#configfile <- "../inst/extdata/config_cyp1b1_zoom_4comet.txt" 
chrom <- "chr2"
start <- 38290160
end <- 38303219
gen <- "hg19"
strant <- "*"

BROWSER.SESSION="UCSC"
mySession <- browserSession(BROWSER.SESSION)
genome(mySession) <- gen

genetrack <-genesENSEMBL(gen,chrom,start,end,showId=FALSE)
snptrack <- snpBiomart(chrom, start, end, dataset="hsapiens_snp_som",showId=FALSE)
strutrack <- structureBiomart(chrom, start, end, strand,
                              dataset="hsapiens_structvar_som",showId=FALSE)
iscatrack <-ISCATrack(gen,chrom,start,end,mySession, table="iscaPathogenic")

listgviz <- list(genetrack,snptrack,iscatrack)
comet(config.file=configfile,TRACKS.GVIZ=listgviz, 
      VERBOSE=FALSE, PRINT.IMAGE=FALSE)
@
\caption{Plot with comet function.\label{fig:cometPlot}}
\end{figure}

\subsection{coMET plot: annotation tracks and correlation matrice}  
It is possible to visualise only annotation tracks and the correlation between genetic elements.
In this case, we need to use the option \texttt{DISP.PVALUEPLOT=FALSE}, for example the figure \ref{fig:cometPlot_nopval}.

<<cometPlotNopvalText, eval=FALSE, fig.keep='last' >>=
library(Gviz)
extdata <- system.file("extdata", package="coMET",mustWork=TRUE)
configfile <- file.path(extdata, "config_cyp1b1_zoom_4comet.txt")
#configfile <- "../inst/extdata/config_cyp1b1_zoom_4comet.txt" 
chrom <- "chr2"
start <- 38290160
end <- 38303219
gen <- "hg19"

genetrack <-genesENSEMBL(gen,chrom,start,end,showId=FALSE)
snptrack <- snpBiomart(chrom, start, end, 
                       dataset="hsapiens_snp_som",showId=FALSE)
strutrack <- structureBiomart(chrom, start, end, 
                              strand, dataset="hsapiens_structvar_som")
clinVariant<-ClinVarMainTrack(gen,chrom,start,end)
clinCNV<-ClinVarCnvTrack(gen,chrom,start,end)
gwastrack <-GWASTrack(gen,chrom,start,end)
geneRtrack <-GeneReviewsTrack(gen,chrom,start,end)

listgviz <- list(genetrack,snptrack,strutrack,clinVariant,
                 clinCNV,gwastrack,geneRtrack)
comet(config.file=configfile,TRACKS.GVIZ=listgviz, 
      VERBOSE=FALSE, PRINT.IMAGE=FALSE,DISP.PVALUEPLOT=FALSE)
@

\begin{figure}
<<cometPlot_nopval, echo=FALSE, fig.keep='last' >>=
library(Gviz)
extdata <- system.file("extdata", package="coMET",mustWork=TRUE)
configfile <- file.path(extdata, "config_cyp1b1_zoom_4comet.txt")
#configfile <- "../inst/extdata/config_cyp1b1_zoom_4comet.txt" 
chrom <- "chr2"
start <- 38290160
end <- 38303219
gen <- "hg19"

genetrack <-genesENSEMBL(gen,chrom,start,end,showId=FALSE)
snptrack <- snpBiomart(chrom, start, end, 
                       dataset="hsapiens_snp_som",showId=FALSE)
strutrack <- structureBiomart(chrom, start, end, 
                              strand, dataset="hsapiens_structvar_som")
clinVariant<-ClinVarMainTrack(gen,chrom,start,end)
clinCNV<-ClinVarCnvTrack(gen,chrom,start,end)
gwastrack <-GWASTrack(gen,chrom,start,end)
geneRtrack <-GeneReviewsTrack(gen,chrom,start,end)

listgviz <- list(genetrack,snptrack,strutrack,clinVariant,
                 clinCNV,gwastrack,geneRtrack)
comet(config.file=configfile,TRACKS.GVIZ=listgviz, 
      VERBOSE=FALSE, PRINT.IMAGE=FALSE,DISP.PVALUEPLOT=FALSE)
@
\caption{Plot with comet function without pvalue plot.\label{fig:cometPlot_nopval}}
\end{figure}


\clearpage
\section*{SessionInfo}
The following is the session info that generated this vignette:
<<session-info, results="asis">>=
toLatex(sessionInfo())
@
\end{document}